lib/matplotlib/widgets.py

```diff
@@ -706,77 +704,60 @@ def __init__(self, ax, label, initial='',
 
         self.DIST_FROM_LEFT = .05
 
-        self.text = initial
-        self.label = ax.text(-label_pad, 0.5, label,
-                             verticalalignment='center',
-                             horizontalalignment='right',
-                             transform=ax.transAxes)
-        self.text_disp = self._make_text_disp(self.text)
+        self.label = ax.text(
+            -label_pad, 0.5, label, transform=ax.transAxes,
+            verticalalignment='center', horizontalalignment='right')
+        self.text_disp = self.ax.text(
+            self.DIST_FROM_LEFT, 0.5, initial, transform=self.ax.transAxes,
+            verticalalignment='center', horizontalalignment='left')
 
         self.cnt = 0
         self.change_observers = {}
         self.submit_observers = {}
 
-        # If these lines are removed, the cursor won't appear the first
-        # time the box is clicked:
-        self.ax.set_xlim(0, 1)
-        self.ax.set_ylim(0, 1)
+        ax.set(
+            xlim=(0, 1), ylim=(0, 1),  # s.t. cursor appears from first click.
+            navigate=False, facecolor=color,
+            xticks=[], yticks=[])
 
         self.cursor_index = 0
 
-        # Because this is initialized, _render_cursor
-        # can assume that cursor exists.
-        self.cursor = self.ax.vlines(0, 0, 0)
-        self.cursor.set_visible(False)
+        self.cursor = ax.vlines(0, 0, 0, visible=False,
+                                transform=mpl.transforms.IdentityTransform())
 
         self.connect_event('button_press_event', self._click)
         self.connect_event('button_release_event', self._release)
         self.connect_event('motion_notify_event', self._motion)
         self.connect_event('key_press_event', self._keypress)
         self.connect_event('resize_event', self._resize)
-        ax.set_navigate(False)
-        ax.set_facecolor(color)
-        ax.set_xticks([])
-        ax.set_yticks([])
+
         self.color = color
         self.hovercolor = hovercolor
 
-        self._lastcolor = color
-
         self.capturekeystrokes = False
 
-    def _make_text_disp(self, string):
-        return self.ax.text(self.DIST_FROM_LEFT, 0.5, string,
-                            verticalalignment='center',
-                            horizontalalignment='left',
-                            transform=self.ax.transAxes)
+    @property
+    def text(self):
+        return self.text_disp.get_text()
 
     def _rendercursor(self):
         # this is a hack to figure out where the cursor should go.
         # we draw the text up to where the cursor should go, measure
         # and save its dimensions, draw the real text, then put the cursor
         # at the saved dimensions
 
-        widthtext = self.text[:self.cursor_index]
-        no_text = False
+        text = self.text
+        widthtext = text[:self.cursor_index]
         if widthtext in ["", " ", "  "]:
-            no_text = widthtext == ""
             widthtext = ","
```

Comment:
This is broken for putting the cursor in the first position, because it ends up in the second position, or more likely, on top of the first character (since they're usually wider than `,`).

Reply:
This seems to work:
```suggestion
            widthtext = "\N{Zero Width Space}"
```

Reply:
Good catch; went for something closer to the original solution (force x1 = x0) just in case some fonts don't have zero-width space.  There's still the problem that spaces at the end get dropped, but this was already present before and a general problem anyways...